### ISSUE TITLE
[HOLD OFFLINE] [$250] Workspace - Same user is added multiple times when tapping invite member CTA multiple times in offline

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#6676]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Precondition: Kill the internet connection
1. Navigate to the Workspace editor
2. Navigate back to the Manage members page
3. Click on invite and enter a user address (you can access to)
4. Select the member and click on invite
5. Tap the invite button multiple times
6. Tap on back button and check added user

## Expected Result:
User should be added only 1 time.

## Actual Result:
Able to add same user many times when offline.
And then selecting one member selects all the members with same name and email

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:**  v1.1.19-0
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:**
![Bug5366090_Screen_Shot_2021-12-09_at_12 33 58_PM](https://user-images.githubusercontent.com/43996225/145459931-702708f8-84

https://user-images.githubusercontent.com/43996225/145459963-e6699ae4-3c3b-4c04-9b43-2e4f0cf52e4f.mp4

https://user-images.githubusercontent.com/43996225/145459951-0624aedb-8941-4f3e-afd6-690298484e52.mp4

b7-4141-a474-93ebf7b93fed.png)

**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause
**Slack conversation:** 
**Upwork URL:** https://www.upwork.com/jobs/~01d54d4c2a0f4e6417

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @pecanoro (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by pecanoro** [comment]
@parasharrajat Is this one also related to our offline issues? Not sure if we should leave this one open if the fix will be the same.

**Comment by parasharrajat** [comment]
I think we can prevent the duplicates without too hassle here. According to the way this component is configured, it should filter exisitng members from the invite list. So when add a new user Policy data trigger a prop update to this component but we are not handling prop changes. 
Even if user is online, He can click the button multiple times and then duplicate entries will be created but After APi call is successful, we remove all the duplicate entries so It is not visible to the user when online.

In case of offline, API call is never completed thus it is apparant. I think it is still necessary to rerender the component when members in the policy are added so that already added user will be removed from the UI.

The End effect would be that When user presses the Invite button. the user from the list will vanish. If we talk about the video, the only searched user will be remvoed from the search results as soon as it is added to the policy. I think it does not harm in any way. 

but Again the fact that invited users are added to the list on the device but not on the backend while offline comes under the offline issues you mentioned Which is not part of this issue anyway.

**Comment by pecanoro** [comment]
Cool, thanks for the insight, so many offline issues lately that I am getting all of them mixed up! I am going to add the external label then!

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by parasharrajat** [comment]
@pecanoro but I am not able to determine what should be done in this issue. What would be the expected behaviour here?

When the user has selected a contact. Should it be removed when added to the policy? or Should we show the loader on the button when an invite call is made? 

I think we should first set clear expectations before moving it to external.

**Comment by pecanoro** [comment]
I am not sure if I am following. If we try to add the same user a second time, we should simply not allow the user to do it, right? Or maybe we should simply not allow inviting member when we are offline.

**Comment by parasharrajat** [comment]
Aha, got you.
I was saying that should we remove that user from the list. For example, you searched for a user and select one, and press invite. Then remove that user from the list. To continue will be disabled automatically. But it may not resolve the issue completely. user can press quickly to add the same user again.

But you are saying to just filter the invited users before adding them so that duplicates are not added. In that case, it would make more sense to show the error for the duplicate users.

**Comment by MelvinBot** [comment]
@trjExpensify Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by trjExpensify** [comment]
Sorry guys, I'm a tad unclear on the next steps here?

**Comment by pecanoro** [comment]
@trjExpensify Export the issue, we can discuss it in the meantime.

@parasharrajat Oh I see, we should remove the user from the list then after being invited, right? I think that would be the best option.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NikkiWines (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by trjExpensify** [comment]
>@trjExpensify Export the issue, we can discuss it in the meantime. 

Done, here: https://www.upwork.com/jobs/~01a872960138b42dec

**Comment by songlang1994** [comment]
I think we should consider the duplicated items of employeeList in `Policy#invite` instead of just simply concat the employeeList.

https://github.com/Expensify/App/blob/893762107a3691a79256db064181a369f3c50384/src/libs/actions/Policy.js#L287-L294

Both `WorkspaceMembersPage` and `WorkspaceInvitePage` depend on the "correct" `employeeList`.

https://github.com/Expensify/App/blob/893762107a3691a79256db064181a369f3c50384/src/pages/workspace/WorkspaceInvitePage.js#L92-L95

https://github.com/Expensify/App/blob/893762107a3691a79256db064181a369f3c50384/src/pages/workspace/WorkspaceMembersPage.js#L246-L252

**Proposal**

```diff
- policy.employeeList = [...policy.employeeList, ...newEmployeeList];
+ policy.employeeList = _.uniq([...policy.employeeList, ...newEmployeeList], employee => employee.login);
```

Uniq the employeeList is not only make sense for the UI rendering, but also makes the API parameters more reasonable.

Limitation:
1. This implementation may have a performance issue since `_.uniq` is an O(n^2) operation.
2. It still results in multiple times API call when the user clicks the Invite button quickly.

**Comment by parasharrajat** [comment]
Thanks for the proposal @songlang1994 .
Based on this.
1 Your proposal satisfies first. We don't need to worry about the complexity here. I don't think a group will ever have thousands of users.
2. Can you think about the second?

**Comment by songlang1994** [comment]
@parasharrajat `Policy#invite` itself is not easy to do this kind of restriction. However, we can handle it from its caller. Like:

https://github.com/Expensify/App/blob/893762107a3691a79256db064181a369f3c50384/src/pages/workspace/WorkspaceInvitePage.js#L61-L65

```diff
- this.inviteUser = this.inviteUser.bind(this);
+ this.inviteUser = _.debounce(this.inviteUser.bind(this), 1000, true)
}
```

**Comment by parasharrajat** [comment]
I think this is not what we want. We want the selected user to be removed from the list when it is added to the policy. let me know if you have any questions.

**Comment by songlang1994** [comment]
> We want the selected user to be removed from the list when it is added to the policy.

This way may not be what the users expected and may cause misunderstandings.

In the offline scenario, after a user clicks the "Invite" Button, the selected items are removed from the list immediately. The user would expect the operation to be completed at this time, but in fact, it has not. After a few seconds, the app display an error message saying that: sry, the operation was not completed. Therefore the user will have to repeat the steps he has already done again.

**Comment by K4tsuki** [comment]
#### proposal

Summary:

My proposal is to keep the UX similar to when the App is online. After clicking `Invite` on `WorkspaceInvitePage`, App will immediately `goback` to `WorkspaceMembersPage`.  App will diplay additional info on member row with unfinished invite request, something similar to : 

<details>
<summary>
Example:
</summary>

The text could be `Sending invitation, waiting for connection...`

![addtional_info](https://user-images.githubusercontent.com/1313124/147574879-016b01a8-942f-4c3e-949f-131ca0d17115.png)
or
![addtional_info_2](https://user-images.githubusercontent.com/1313124/147574889-3b3c6184-c5ab-4e20-97b7-771a178fd854.png)

</details>

__________________
Details:

The App needs additional field on already exists onyx policy data. Additional field could be called `unfinishedInviteRequest` that contains emails of unfinished invite API request.

<details> <summary> Adtional field details: </summary>
The data could be set before this onyx merge in `Policy.invite` function:

https://github.com/Expensify/App/blob/149ed21c36839a53ab4ce16c4b5b505ef792a2e6/src/libs/actions/Policy.js#L300

And on `then` clause 

https://github.com/Expensify/App/blob/149ed21c36839a53ab4ce16c4b5b505ef792a2e6/src/libs/actions/Policy.js#L308

App will remove `unfinishedInviteRequest` emails when API request  is  successful.
On offline state, the `unfinishedInviteRequest` will remains and member page will display info similar to image above.

When reconnecting / back to online:
On accessing `WorkspaceMemberPage` again, `loadFullPolicy` should already update policy data and `unfinishedInviteRequest`. I think this function should be called each time `WorkspaceInitialPage` `didMounted`, to sync  policy data  between server and local data.

</details>


<br />
<details>
<summary >##### prevent double click on invite button (prevent multiple invitation API request and multiple member displayed in invite user list (online and offline)): </summary>

By adding a property to `WorkspaceInvitePage` for example: `isSubmitActive` with default to true and check this property in this function: 

https://github.com/Expensify/App/blob/149ed21c36839a53ab4ce16c4b5b505ef792a2e6/src/pages/workspace/WorkspaceInvitePage.js#L211

setting the `isSubmitActive` to false after `isSubmitActive` check.

</details>

<br />
<details>
<summary >##### loadFullPolicy bug, I think </summary>

This function isn't updating workspace members correctly, I think this line should use `Onyx.set`not  `Onyx.merge`. And policy data should be adjusted for `Onyx.set`

https://github.com/Expensify/App/blob/149ed21c36839a53ab4ce16c4b5b505ef792a2e6/src/libs/actions/Policy.js#L226

</details>

**Comment by parasharrajat** [comment]
Interesting thoughts @K4tsuki. Could you please initiate a slack convo as you are suggesting the UI change?

**Comment by developvsn** [comment]
> Interesting thoughts @K4tsuki. Could you please initiate a slack convo as you are suggesting the UI change?

When I was working on multiple tasks, You pointed that I am a new contributor and suggested to check contributors guide [here](https://github.com/Expensify/App/issues/6386#issuecomment-975413724). 

This time I would like to point that You're not objecitve, and have double standards because you are not suggesting @k4tsuki to check the contributors guide the same you did to us.

**Comment by K4tsuki** [comment]
> Maybe you have a personal interest in hiring this user?

This is not true, I have been warned too, In [here](https://github.com/Expensify/App/issues/5699#issuecomment-998061239) and [here](https://github.com/Expensify/App/issues/6474#issuecomment-997803346). But my pull request has been accepted in [here](https://github.com/Expensify/App/pull/6849#pullrequestreview-840542102). <br />
I also submit complaint in [here](https://github.com/Expensify/App/issues/6759#issuecomment-996344792) too, because I think the accepted proposal from other user is similar to my original proposal.  <br />
My proposal was not accepted in [here](https://github.com/Expensify/App/issues/5699#issuecomment-997573670) too, because paraharrajat already take the issue, I think. <br />
If you take close look at proposal acceptance  in [here](https://github.com/Expensify/App/issues/6563#issuecomment-996347140), I kept pushing forward to get my proposal accepted. <br />

**Comment by parasharrajat** [comment]
Thanks, both of you for showing your concerns. 

@developvsn, I haven't selected his proposal, or did I? @K4tsuki has suggested a few UI changes that are not necessarily needed for this issue but could be useful. so I asked him to raise this in Slack.

Let me know if still feel the same. I would be happy to answer your questions. You can always raise your concerns in slack.

**Comment by K4tsuki** [comment]
slack : https://app.slack.com/client/T03SC9DTT/C01GTK53T8Q/thread/C01GTK53T8Q-1640763879.205800

**Comment by NikkiWines** [comment]
Hmm, I wonder if we might just re-use the `You appear to be offline` notification and prevent the user from inviting those users until they're back online.

![Screen Shot 2022-01-03 at 17 36 11](https://user-images.githubusercontent.com/16747822/147987806-d4b84ff1-9b1f-4e24-8f01-98730885efbd.png)

However, if we want to go the route of still allowing users to go through that action I agree we should further discuss the UI changes. I'll bump the thread in slack.

**Comment by NikkiWines** [comment]
An alternative UI change might be to grey-out pending members and update the offline message to indicate that pending members will be added later.

![Screen Shot 2022-01-03 at 18 00 02](https://user-images.githubusercontent.com/16747822/147989524-b0be608f-820c-4c3b-ae43-a0a8e0acc140.png)

 
cc: @Expensify/design you might have some opinions on how to handle this. Relevant slack conversation [here](https://app.slack.com/client/T03SC9DTT/C01GTK53T8Q/thread/C01GTK53T8Q-1640763879.205800) if you'd rather discuss there.

**Comment by shawnborton** [comment]
I like the approach you outlined above.

**Comment by parasharrajat** [comment]
The plan sounds good but we have a problem. We don't queue the requests for invitations. So when users will be online no further requests will be sent for invited users. 

This could be an indication that we should hold this issue like the other offline issues.

Also, we need to store the pending invitation information in the policy data. So this issue requires planning.

**Comment by K4tsuki** [comment]
> 

Actually the requests is queued. What do you mean by not queued? Even if user close the tabs or browser when they are offline, there is already `{persist: true}` facility to persist the queue in local storage.

**Comment by parasharrajat** [comment]
Aha, @K4tsuki Could you please show me where MemberInvite requests are being persisted? I may have overlooked it.

**Comment by K4tsuki** [comment]
In this line we could persist API request by:

https://github.com/Expensify/App/blob/691dda2c853394a2c8f86dba55a5680842daf516/src/libs/Network.js#L181-L183

adding parameter `{persist: true}` on API request data.
<br />
For invite member, could be add in here:

https://github.com/Expensify/App/blob/691dda2c853394a2c8f86dba55a5680842daf516/src/libs/API.js#L907-L913

**Comment by parasharrajat** [comment]
> We don't queue the requests for invitations https://github.com/Expensify/App/issues/6676#issuecomment-1004739075

Exactly, it's not there and we might have to add that.

**Comment by trjExpensify** [comment]
How are we getting on here, guys? Are we moving forward with @K4tsuki proposal?

**Comment by parasharrajat** [comment]
Not yet. I think Nikkiwines raised a slack discussion and thus we are waiting for the conclusion.

**Comment by NikkiWines** [comment]
Oh sorry for the confusion - I didn't end up bumping the slack thread @K4tsuki made [above](https://github.com/Expensify/App/issues/6676#issuecomment-1002463361) because I ended up cc'ing the design team in my comment [here](https://github.com/Expensify/App/issues/6676#issuecomment-1004412810). Since @shawnborton agreed with my UI suggestion I figured the only outstanding comments was @parasharrajat's comment [here](https://github.com/Expensify/App/issues/6676#issuecomment-1004923658) which has yet to be addressed.

**Comment by parasharrajat** [comment]
Ok. Thanks, @NikkiWines  for clarifying. 
@K4tsuki Could you please create a proposal to achieve https://github.com/Expensify/App/issues/6676#issuecomment-1004412810 and all other necessary pieces of the puzzle?

**Comment by K4tsuki** [comment]
All right, please wait.

**Comment by K4tsuki** [comment]
This is my plan:

<details> <summary>For offline text component: </summary>

I will add `OfflineText` component, that will check offline status and display offline message. [The properties might be: `layoutStyle`, `textStyle`, `additonalText`, etc ]

</details>

<details> <summary>For `option row` styling </summary>

Add `pendingInvitation` properties to policy onyx data that will contain list of pending invitation emails.
Add `pending` properties in `data` option for the pending invitations.  

https://github.com/Expensify/App/blob/7d8d762a75af633eb5bbb9a8a5008ced1705dffa/src/pages/workspace/WorkspaceMembersPage.js#L249

Add display text styling logic for `pending` option in sidebar `OptionRow`.

https://github.com/Expensify/App/blob/7d8d762a75af633eb5bbb9a8a5008ced1705dffa/src/pages/home/sidebar/OptionRow.js#L85-L90

</details>

<details> <summary>API change: </summary>

Add `{persist: true}` in `Policy_Employees_Merge` API post parameter, to make it persistent and will be stored in Onyx local data when `offline`.

https://github.com/Expensify/App/blob/691dda2c853394a2c8f86dba55a5680842daf516/src/libs/API.js#L907-L913

</details>

<details> <summary>Data Flow  </summary>

Update onyx value of `pendingInvitation` in:
1. `loadFullPolicy`. In here, check if `employeeList` contains `pendingInvitation` email item. If found, then remove `pendingInvitation` email.
https://github.com/Expensify/App/blob/ded47c96d033a510bd174e11ba917cdeff1ca5d6/src/libs/actions/Policy.js#L215

2. In each success on `invite` user and `remove` member, we update `pendingInvitation` and `empolyeeList`accordingly.

3. In `invite` function below. We add invited emails to `pendingInvitation` instead of `employeeList`. After merge, app will goBack immediately.
https://github.com/Expensify/App/blob/149ed21c36839a53ab4ce16c4b5b505ef792a2e6/src/libs/actions/Policy.js#L300
then when request success (in case of online), remove this emails in `pendingInvitation` and add it in `employeeList`. In offline, `pendingInvitation` keep remain.

4. If user remove `pending` invitation user. App will check onyx data for `networkRequestQueue` and delete Api request for that invitation. Also remove onyx data on `pendingInvitation` in policydata.


</details>

<details> <summary>Prevent double click on invite user: </summary>

By adding a property to `WorkspaceInvitePage` for example: `submitActive` with default to true and check this property on `inviteUser`, disable `submitActive` immediately after passing the check.

https://github.com/Expensify/App/blob/149ed21c36839a53ab4ce16c4b5b505ef792a2e6/src/pages/workspace/WorkspaceInvitePage.js#L211

</details>

More tests must be done, then adjust the code accordingly.

**Comment by parasharrajat** [comment]
> For `option row` styling 

Better to just pass the styles from props instead of adding new props. 

> Prevent double click on invite user: 

We don't want to block the button in this case. It won't look good. The button should only be disabled when no user is selected. There should simply be no effect for inviting the same user again. It means you will add a check on the action so that it ignores duplicate invites.

> pendingInvitation

We first need to modify the API on the backend to allow storing this value.

**Comment by parasharrajat** [comment]
@K4tsuki's proposal looks good to me. Do you have any questions?
The plan could be:
1. @K4tsuki submits the PR for UI change. 
2. We keep this PR on hold until Backend changes are done. 
3. Test the PR when the backend is ready. 
4. Merge it. 


cc: @NikkiWines 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by K4tsuki** [comment]
> We first need to modify the API on the backend to allow storing this value.

My proposal is on Onyx local data. There is no need to modify API/backend.

> We don't want to block the button in this case. It won't look good. 

I am not blocking it. The style / display of button is same  (no change).
After click `invite` button, App will immediately go back to member page. 
It will disable multiple clicking on invite button, because  the function is just single click on `invite` button and the goback. It just make sure that feature/function run the way user assume it.

**Comment by parasharrajat** [comment]
> My proposal is on Onyx local data. There is no need to modify API/backend.

How will you pass pending invitation to other devices?

**Comment by K4tsuki** [comment]
It is for offline mode purpose. so it is for specific device that goes offline.
Like when you sending message using device that goes offline. The message will be send from that device if it goes online again.

**Comment by parasharrajat** [comment]
Yeah, you are correct. I forgot we are offline :rofl: So when user will be online, pending invitation will automatically be removed.

**Comment by K4tsuki** [comment]
In my proposal, the time `pendingInvitation` and `employeeList`  is synced is when user open workspace, user add member and user remove member. 

It would be best to make procedure that subscribe / connect to `NetworkRequestQueue` onyx data and that procedure will call a callback based on what queued API request is being send and provide response of the request to the callback. 

For example on this problem the procedure that subscribe /connect to `NetworkRequestQueue` will call a callback that will update `workspace member` and `pending invitation` after sending  persisted queued API `Policy_Employees_Merge`.
This could be used by other APIs for offline purpose. 
But maybe this is a big change and I don't know if I have permission to do this.

**Comment by parasharrajat** [comment]
Every API call keeps resolve and reject callback. So until the user force closes the App, these callbacks are fired. Thus we don't need any other way of handling this. When requests are saved to persistent storage,  we can't save the callbacks and there is no way of doing that So we have to rely on user refreshing the data by calling the API which would be the case most of the time.

**Comment by K4tsuki** [comment]
All right.

**Comment by trjExpensify** [comment]
@NikkiWines are you good with the proposed solution? @K4tsuki can you apply for the job on Upwork in the meantime [here](https://www.upwork.com/jobs/~01d54d4c2a0f4e6417), please.

**Comment by K4tsuki** [comment]
@trjExpensify thank you, I have applied to the upwork job.

**Comment by NikkiWines** [comment]
Oh, sorry, I missed @parasharrajat's comment above and thought things were still in discussion. @K4tsuki proposal from [here](https://github.com/Expensify/App/issues/6676#issuecomment-1013597261) looks good

**Comment by trjExpensify** [comment]
Cool, sent the offer @K4tsuki. @parasharrajat, likewise for C+.

**Comment by MelvinBot** [comment]
📣 @K4tsuki You have been assigned to this job by @trjExpensify!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by K4tsuki** [comment]
Update: WIP, not done yet.

**Comment by iwiznia** [comment]
@parasharrajat asked me for a spanish translation of `You appear to be offline. Pending invitations will send when you are online.`, but I am not convinced that `pending invitations` is the right copy for this. I would use something like: `You appear to be offline. The people you have invited will be added to the workspace when you are online.`. Thoughts @Expensify/marketing?

**Comment by jamesdeanexpensify** [comment]
Maybe we could say (slight tweaks):

> Looks like you're offline! When you're back online, the people you invited will be added to the workspace.

**Comment by iwiznia** [comment]
I like @jamesdeanexpensify's version a bit better, though I would remove the `!` from it, it's not that alarming that the request will be sent later 😉 

If we are going with that one, the spanish version would be:
> Parece que estás desconectado. La gente que invitaste al espacio de trabajo será agregada cuando te vuelvas a conectar.

**Comment by parasharrajat** [comment]
cc @K4tsuki Please update those as per :arrow_up:

**Comment by trjExpensify** [comment]
Where are we at with the changes requested here to get the PR merged? 

>Looks like you're offline! When you're back online, the people you invited will be added to the workspace.

On this copy @jamesdeanexpensify, did we opt for `people` over `members` for a particular reason? We refer to users on a workspace as `members` (The pages are called `Manage members` and `Invite new members`), so wouldn't it make sense to also use `members` here?

**Comment by mallenexpensify** [comment]
Assigned myself to manage. @K4tsuki has been paid. 
@NikkiWines, @parasharrajat would either of you want to work on this issue or should we reopen it to contributors?

**Comment by parasharrajat** [comment]
I will review the state of PR and let you know if I have the capacity to do it.

**Comment by NikkiWines** [comment]
I believe we're waiting on an update in https://github.com/Expensify/App/pull/7595. When it's updated I'm happy to review, but if we're fully re-assigning the issue I, unfortunately, don't have the bandwidth to take on this issue right now.

**Comment by parasharrajat** [comment]
I won't be able to work on it by next weekend. @mallenexpensify.

**Comment by mallenexpensify** [comment]
@marcaaron, weird coincidence you got auto-assigned this, want to put on hold while you/others work on offline-first? If so I can preface the title. 
WOW.. you just did it while I was typing.  Great minds....

**Comment by marcaaron** [comment]
Let's not do anything related to offline yet.

**Comment by michaelhaxhiu** [comment]
Let's close this for now, as it was opened by Applause and this will be different after we release offline first.
